### ISSUE TITLE
Separate common package

### ISSUE THREAD

**Issue by sagikazarmark** [php-http/common#1]
Actually I am always afraid of packages like "common". It suggests they are only container of everything else. While reusability is a good think, it does not always makes sense. Currently we have the following commonish packages:

- common (traits for interfaces mostly)
- adapter-common (contains implementations of http adapter interfaces)
- adapter-core (deprecated, logic should be moved to adapter-common and adapter-client)
- client-common (does not exist, yet)

This mess should absolutely be cleared up.

/cc @ddeboer

**Comment by sagikazarmark** [comment]
Adapter required common resources:
- Exceptions

Client required common resources:
- Configurable trait impl
- Parameterable message trait impl

Both required common resources:
- Message factory implementations and guesser

**Comment by ddeboer** [comment]
Yes, if possible get rid of the common packages. 

> Exceptions

So why not just move these into the adapter package? What about the adapter testing setup, that's common as well, isn't it?

> Client required common resources:

What do you mean exactly? The library will only ship with one client, right?

> Message factory implementations and guesser

I think a client should only depend on a MessageFactory interface, which goes in the message-factory package, just like the Guesser.

**Comment by sagikazarmark** [comment]
> > Exceptions

> So why not just move these into the adapter package? What about the adapter testing setup, that's common as well, isn't it?

My intention was that the adapter package has no implementations in it

> What do you mean exactly? The library will only ship with one client, right?

In theory, yes. However I would like to separate the client interfaces from the client package. But probably all client related resource should go into the only implementation package.

> I think a client should only depend on a MessageFactory interface, which goes in the message-factory package, just like the Guesser. 

The message factory package is an interface package for PSR-7. I created it for PSR-7. However I am also thinking about that the message factory implementations could go into their own package, because they are going to be used by both the adapters and the client.

**Comment by ddeboer** [comment]
> My intention was that the adapter package has no implementations in it

I don’t see why not. I agree that fully-fledged objects are a bad idea in a stable base package such as adapter, but Exceptions should be relatively stable.

**Comment by sagikazarmark** [comment]
Yeah, probably you are right. I was also thinking about this.

How should I move exceptions back? Remove the interfaces they implement? They does not make much sense actually.

**Comment by ddeboer** [comment]
Haha, you run into this problem because of the lack of the `Interface` suffix. :wink:

Yes, I’d just remove the interfaces, but keep the base `Exception` interface so users can typehint on that.

**Comment by sagikazarmark** [comment]
Well, not exactly.

You are right, but when I removed them from the contract package I realized I still need something, since the interfaces declare to throw those types of exceptions, that's why I added interfaces.

Do you think I have to test those exceptions? :wink:

The adapter testing is an interesting question. It definitely needs to go into a separate package, because it is only needed in development. One possible question if the adapter and client integration tests should go into one package. Currently I removed all client related tests from the test suit. When the client gets ready, they should go somewhere as well. Maybe we could rename adapter-integration-tests to integration-tests?

**Comment by sagikazarmark** [comment]
message-factory-common or message-factory-guesser?

I feel the latter a bit inappropriate since some default implementations are also there.

**Comment by sagikazarmark** [comment]
Shouldn't the separate message factories go into separate packages?

**Comment by ddeboer** [comment]
> One possible question if the adapter and client integration tests should go into one package.

For now, I’d focus on finishing the adapter layer so only the packages required for the adapters. When that works, tag it, and only then move on to the client layer. It’s not so easy to do everything together at the same time. Of course, if working on the client layer requires changes in the adapter packages, that’s fine as long as we’re < 1.0

**Comment by sagikazarmark** [comment]
> We don’t need to worry about that for the adapters, or do we?

Non PSR-7 clients need a MessageFactory instance to create requests/responses.

**Comment by ddeboer** [comment]
Yes, but that message factory will be different for each adapter so can just be included in the buzz-adapter (etc.) package.

**Comment by sagikazarmark** [comment]
Why? Message factories are needed for PSR-7 messages, not the underlying clients. In the adapter the client's message format is fixed, but you can use any kind of PSR-7 message implementations. In case of Guzzle 6 it is not a problem. But in other cases, you have to pick one.

**Comment by ddeboer** [comment]
What I meant was: the code calling the MessageFactory will be specific to each adapter. But you’re completely right the MessageFactory itself will not. So will php-http ship with one default MessageFactory (e.g., guzzle/psr-7 or zend/diactoros) in a separate package?

**Comment by sagikazarmark** [comment]
Currently they are in the common package. There is also a message factory guesser which guesses the message factory based on the installed PSR implementation.

So we have two options here:

1. Collect all message factory implementations for "common" PSR-7 implementations in one package with the guesser.
2. Place each message factory implementation in a separate package (optionally providing a message factory implementation virtual package.

The most ideal would be if all implementations would also provide a message factory implementation using the implementing interface, but that is not possible.

The current content of the common message is going to be separated: the message factory part goes into one (or more?) package, the rest in the client package. Adapter common (exceptions) is moved back to the adapter package.

**Comment by sagikazarmark** [comment]
https://github.com/php-http/message-factory-guesser
